### ISSUE TITLE
pausing stubs/spies within a method

### ISSUE THREAD

**Issue by mmahalwy** [sinonjs/sinon#1562]
tldr; is there a way to pause call counts for stubs and spies temporarily?

Given a method (and specifically, trying to test usage of `mapStateToProps` with react-redux) that returns an object, we'd like to test whether the returned object is equal to when the function is called again. Current solution:

```js
const mapStateToProps = () => {
  const badObject = callSomeMethod();

  return {
    bad: 1,
    badObject: badObject
  };
}
```
Then spec:
```js
const result1 = mapStateToProps(someState);
const result2 = mapStateToProps(someState);

expect(areYouEqual(result1, result2).to.equal(false);
```

This test is correct! Since they will result to non-shallow equal objects. The problem is when specs stub `callSomeMethod`, the call count ends up being 2, as opposed to the intended 1. 

Reseting the store between the calls ends up reseting stubs that may have been called outside of `mapStateToProps`. 
```js
const result1 = mapStateToProps(someState);
sinon.reset();
const result2 = mapStateToProps(someState);

expect(areYouEqual(result1, result2).to.equal(false);
```

Tried a solution such as:

```js
const stubs = sinon.fakes;
const result1 = mapStateToProps(someState);
sinon.fakes = stubs;
const result2 = mapStateToProps(someState);

expect(areYouEqual(result1, result2).to.equal(false);
```
Not great, hacky and may result in inconsistencies. 

Potential/proposed solutions:
```js

sinon.pause();
const result1 = mapStateToProps(someState);
sinon.resume();
const result2 = mapStateToProps(someState);

expect(areYouEqual(result1, result2).to.equal(false);
```
or 
```js
const result1 = sinon.cancelWithin(mapStateToProps(someState));
const result2 = mapStateToProps(someState);

expect(areYouEqual(result1, result2).to.equal(false);
```

Thoughts? Ideas? Can this be done? Too hacky? Need more explanations?

**Comment by mroderick** [comment]
> tldr; is there a way to pause call counts for stubs and spies temporarily?

No. Doing so would seem like a bit of an anti-pattern to me.

> Since they will result to non-shallow equal objects. The problem is when specs stub callSomeMethod, the call count ends up being 2, as opposed to the intended 1.

Why would you expect it to be `1` when you're calling it twice.

**Comment by mmahalwy** [comment]
@mroderick this is more of a global monkey patch to flag to developers when `mapStateToProps` is not used properly. Thus, we are calling it twice to test that. We'd like to not fail existing tests that are testing for the stub being called once, which without the monkey patch is truthy.

**Comment by fatso83** [comment]
:-1: It's quite hard to follow the _implied_ test code, but it seems like it is testing both the assertion we are seeing above, as well as the call count. This is me assuming this, but wouldn't this problem go away by having simpler tests testing _one_ thing at a time? Test the call count is right in one test and test the results in another.

Anyway, no matter my reasoning above, I don't find having this a good idea. It paves the way for more complex tests and will make Sinon more complex internally.

**Comment by mmahalwy** [comment]
Agreed! Going to close this
